config/locals.go

```diff
@@ -0,0 +1,313 @@
+type Local struct {
+	Name string
+	Expr hcl.Expression
+}
+
+// evaluateLocalsBlock is a routine to evaluate the locals block in a way to allow references to other locals. This
+// will:
+// - Extract a reference to the locals block from the parsed file
+// - Continuously evaluate the block until all references are evaluated, defering evaluation of anything that references
+//   other locals until those references are evaluated.
+// This returns a map of the local names to the evaluated expressions (represented as `cty.Value` objects). This will
+// error if there are remaining unevaluated locals after all references that can be evaluated has been evaluated.
+func evaluateLocalsBlock(
+	terragruntOptions *options.TerragruntOptions,
+	parser *hclparse.Parser,
+	hclFile *hcl.File,
+	filename string,
+) (map[string]cty.Value, error) {
+	diagsWriter := util.GetDiagnosticsWriter(parser)
+
+	localsBlock, diags := getLocalsBlock(hclFile)
+	if diags.HasErrors() {
+		diagsWriter.WriteDiagnostics(diags)
+		return nil, errors.WithStackTrace(diags)
+	}
+	if localsBlock == nil {
+		// No locals block referenced in the file
+		terragruntOptions.Logger.Printf("Did not find any locals block: skipping evaluation.")
+		return nil, nil
+	}
+
+	terragruntOptions.Logger.Printf("Found locals block: evaluating the expressions.")
+
+	locals, diags := decodeLocalsBlock(localsBlock)
+	if diags.HasErrors() {
+		terragruntOptions.Logger.Printf("Encountered error while decoding locals block into name expression pairs.")
+		diagsWriter.WriteDiagnostics(diags)
+		return nil, errors.WithStackTrace(diags)
+	}
+
+	// Continuously attempt to evaluate the locals until there are no more locals to evaluate, or we can't evaluate
+	// further.
+	evaluatedLocals := map[string]cty.Value{}
+	evaluated := true
+	for len(locals) > 0 && evaluated {
+		var err error
+		locals, evaluatedLocals, evaluated, err = attemptEvaluateLocals(terragruntOptions, filename, locals, evaluatedLocals)
+		if err != nil {
+			terragruntOptions.Logger.Printf("Encountered error while evaluating locals.")
+			return nil, err
+		}
+	}
+	if len(locals) > 0 {
+		// This is an error because we couldn't evaluate all locals
+		terragruntOptions.Logger.Printf("Not all locals could be evaluated:")
+		for _, local := range locals {
+			terragruntOptions.Logger.Printf("\t- %s", local.Name)
+		}
+		return nil, errors.WithStackTrace(CouldNotEvaluateAllLocalsError{})
+	}
+
+	return evaluatedLocals, nil
+}
+
+// attemptEvaluateLocals attempts to evaluate the locals block given the map of already evaluated locals, replacing
+// references to locals with the previously evaluated values. This will return:
+// - the list of remaining locals that were unevaluated in this attempt
+// - the updated map of evaluated locals after this attempt
+// - whether or not any locals were evaluated in this attempt
+// - any errors from the evaluation
+func attemptEvaluateLocals(
```

Comment:
Yikes, this is quite complicated. I have some vague memory that the HCL parser has native support for partial evaluation specifically for this use case, but can't find it in the code base quickly...

Reply:
I searched but the only partial evaluation I could find was at the top file level, and not at the individual block level where we have expressions.